### PR TITLE
perf(@angular-devkit/build-angular): cache JavaScriptOptimizerPlugin results

### PR DESCRIPTION
**Pull request by alan-agius4**
With this change we cache JavaScriptOptimizerPlugin results in memory or on the file system based on the caching strategy used.

### CODE REVIEW

File: packages/angular_devkit/build_angular/src/webpack/plugins/javascript-optimizer-plugin.ts [hunk 1]
```diff
@@ -84,19 +85,37 @@ export class JavaScriptOptimizerPlugin {
...
+                }
+
+                const { source, map } = scriptAssetSource.sourceAndMap();
                 scriptsToOptimize.push({
                   name: scriptAsset.name,
                   code: typeof source === 'string' ? source : source.toString(),
                   map,
                 });
+
+                cachedItems[name] = cacheItem;
```
Comment (clydin) [inline packages/angular_devkit/build_angular/src/webpack/plugins/javascript-optimizer-plugin.ts:118]: I think `cacheItem` could be stored in the object literal pushed into `scriptsToOptimize` above this line.  This should allow removal of the `cachedItems` object map.

File: packages/angular_devkit/build_angular/src/webpack/plugins/javascript-optimizer-plugin.ts [hunk 2]
```diff
@@ -84,19 +85,37 @@ export class JavaScriptOptimizerPlugin {
           stage: compiler.webpack.Compilation.PROCESS_ASSETS_STAGE_OPTIMIZE_SIZE,
         },
         async (compilationAssets) => {
+          const cache = compilation.getCache('JavaScriptOptimizerPlugin');
```
Comment (clydin) [inline packages/angular_devkit/build_angular/src/webpack/plugins/javascript-optimizer-plugin.ts:88]: Would it be worthwhile to add a check here if caching is enabled and make `cache` possibly `undefined`?  With the potential advantage of skipping the cache related calls below.  I'm not sure if there is an easy and definitive way to determine cache enablement though.

Comment (alan-agius4) [reply packages/angular_devkit/build_angular/src/webpack/plugins/javascript-optimizer-plugin.ts:88]: I think that should be possible via `compilation.options.experiments.cache`.

Comment (alan-agius4) [reply packages/angular_devkit/build_angular/src/webpack/plugins/javascript-optimizer-plugin.ts:88]: Actually, it appears that will other experiments options can be accessed via `compilation.options.experiments`, `cache` is not.

File: packages/angular_devkit/build_angular/src/webpack/plugins/javascript-optimizer-plugin.ts [hunk 3]
```diff
@@ -85,19 +86,42 @@ export class JavaScriptOptimizerPlugin {
         },
         async (compilationAssets) => {
           const scriptsToOptimize = [];
+          const cache = compilation.getCache('JavaScriptOptimizerPlugin');
```
Comment (clydin) [inline packages/angular_devkit/build_angular/src/webpack/plugins/javascript-optimizer-plugin.ts:89]:
```suggestion
          const cache = !!compilation.options.cache && compilation.getCache('JavaScriptOptimizerPlugin');
```
nit: checking here should avoid the `getCache` call if not needed.

**Review by clydin — APPROVED** [review]

File: packages/angular_devkit/build_angular/src/webpack/plugins/javascript-optimizer-plugin.ts (hunk 3 above)
Comment (alan-agius4) [reply packages/angular_devkit/build_angular/src/webpack/plugins/javascript-optimizer-plugin.ts:89]: I was going for that but I did complicate the type in Line 100.

Will try to give it another shoot.

**Pull request merged**